mwan3: mwan3track via default routing table and use procd from mwan3track & mwan3rtmon

aaronjg opened the pull request:
Maintainer:  @feckert 
Compile tested: 19.07.3-ipq806x and snapshot-7673df79 MT76x8 
Run tested: 19.07.3 and snapshot - 7673df79

Description:
This P/R has a few fixes for the 19.07.3

1) Use the default routing tables for all mwan3track outgoing requests. This fixes some issues on the 5.4.x kernel where packets are not routed properly unless they go through a routing table that has the gateway. This also allows us to remove some hacks that were in place for ipv6 mwan3track
2) Comply with source based routing by copying over the default routes from the main routing table rather than creating them anew.
3) Use procd to manage the mwan3rtmon and mwan3track scripts.

This will fix many of the issues with tracking packets originating from the wrong interface or with the wrong IP address. Fixes #8139, fixes #10712, fixes #11406 and fixes #12836, 

Also some other open issues. Fixes #13003 and fixes #13200

aaronjg commented:
Under my testing, with the 5.4.x kernel, if the default policy action for 0.0.0.0/0 is 'unreachable', then when an interface is down mwan3track will not be able to send out pings (or other tracking packets).  It is important that this is merged before the OpenWRT 20.x.x stable release.

openwrt-diy commented:
> To test on 19.x, I recompiled libcap with:
> `CONFIG_PACKAGE_libcap-bin=y`

Error compiling here，packages/net/mwan3/Makefile
     +iptables-mod-extra \
     -libcap-bin \
     +libcap \

openwrt-diy commented:
**All interfaces are offline, unable to access the Internet!**
PKG_NAME:=mwan3
PKG_VERSION:=2.9.0
PKG_RELEASE:=1

![001](https://user-images.githubusercontent.com/57249132/90803805-e313cf00-e34b-11ea-8bc3-450b6b0b217b.png)
![002](https://user-images.githubusercontent.com/57249132/90803814-e73fec80-e34b-11ea-8ab7-6fe413a1c2fd.png)
![003](https://user-images.githubusercontent.com/57249132/90803824-ea3add00-e34b-11ea-87e6-894d8572e90f.png)

openwrt-diy commented:
kokang@virtual-machine:~/openwrt-19.07$ ./scripts/feeds install -a
Collecting package info: done
Collecting target info: done
Installing all packages from feed packages.
Installing all packages from feed luci.
Installing all packages from feed routing.
Installing all packages from feed telephony.

kokang@virtual-machine:~/openwrt-19.07$ make defconfig
tmp/.config-package.in:108142:error: recursive dependency detected!
For a resolution refer to Documentation/kbuild/kconfig-language.txt
subsection "Kconfig recursive dependency limitations"
tmp/.config-package.in:108142:	symbol PACKAGE_iptables is selected by PACKAGE_mwan3
For a resolution refer to Documentation/kbuild/kconfig-language.txt
subsection "Kconfig recursive dependency limitations"
tmp/.config-package.in:114086:	symbol PACKAGE_mwan3 is selected by PACKAGE_luci-app-mwan3
For a resolution refer to Documentation/kbuild/kconfig-language.txt
subsection "Kconfig recursive dependency limitations"
tmp/.config-package.in:73569:	symbol PACKAGE_luci-app-mwan3 depends on PACKAGE_libcap
For a resolution refer to Documentation/kbuild/kconfig-language.txt
subsection "Kconfig recursive dependency limitations"
tmp/.config-package.in:65163:	symbol PACKAGE_libcap is selected by PACKAGE_ip-full
For a resolution refer to Documentation/kbuild/kconfig-language.txt
subsection "Kconfig recursive dependency limitations"
tmp/.config-package.in:113997:	symbol PACKAGE_ip-full is selected by PACKAGE_libreswan
For a resolution refer to Documentation/kbuild/kconfig-language.txt
subsection "Kconfig recursive dependency limitations"
tmp/.config-package.in:127684:	symbol PACKAGE_libreswan depends on PACKAGE_iptables
#
# configuration written to .config
#

openwrt-diy commented:
**An error occurred during compilation, luci-app-mwan3 could not be found**

kokang@virtual-machine:~/openwrt-19.07$ make menuconfig
Collecting package info: done
Collecting target info: done
tmp/.config-package.in:108142:error: recursive dependency detected!
For a resolution refer to Documentation/kbuild/kconfig-language.txt
subsection "Kconfig recursive dependency limitations"
tmp/.config-package.in:108142:	symbol PACKAGE_iptables is selected by PACKAGE_mwan3
For a resolution refer to Documentation/kbuild/kconfig-language.txt
subsection "Kconfig recursive dependency limitations"
tmp/.config-package.in:114086:	symbol PACKAGE_mwan3 is selected by PACKAGE_luci-app-mwan3
For a resolution refer to Documentation/kbuild/kconfig-language.txt
subsection "Kconfig recursive dependency limitations"
tmp/.config-package.in:73569:	symbol PACKAGE_luci-app-mwan3 depends on PACKAGE_libcap
For a resolution refer to Documentation/kbuild/kconfig-language.txt
subsection "Kconfig recursive dependency limitations"
tmp/.config-package.in:65163:	symbol PACKAGE_libcap is selected by PACKAGE_ip-full
For a resolution refer to Documentation/kbuild/kconfig-language.txt
subsection "Kconfig recursive dependency limitations"
tmp/.config-package.in:113997:	symbol PACKAGE_ip-full is selected by PACKAGE_libreswan
For a resolution refer to Documentation/kbuild/kconfig-language.txt
subsection "Kconfig recursive dependency limitations"
tmp/.config-package.in:127684:	symbol PACKAGE_libreswan depends on PACKAGE_iptables


Your configuration changes were NOT saved.

aaronjg commented:
> kokang@virtual-machine:~/openwrt-19.07$ make menuconfig

This requires the libcap-bin package, which is not available on 19.x at this time. See this:

https://github.com/openwrt/packages/pull/13169#issuecomment-676294855

If you want to build the whole 19.07 image, you need to comment out the dependency on the separate libcap-bin package dependency.

jamesmacwhite commented:
@openwrt-diy For quick and dirty testing, to save a recompile, I found you can use the packages from snapshot: https://downloads.openwrt.org/snapshots/packages and install libcap and libcap-bin with opkg manually on 19.07. Make sure to replace libcap as well, otherwise things will be broken. A clean recompile is of course better, but to shortcut testing it does work on 19.07. There doesn't appear to be any kernel dependency so you get away with it.

Unless the libcap-bin change is backported to 19.07, it creates a bit of a problem for having this in 19.07 currently.

openwrt-diy commented:
> @aaronjg Thanks for this!
> 
> I recently experimented with L2TP and it does seem pretty broken currently even under 19.07.x. I believe the commit here to use default routing table helps the situation a lot.
> 
> [1f6c49c](https://github.com/openwrt/packages/commit/1f6c49cf591df6cbfe2538097fb1371f1f8b7ca7)
> 
> I tested out an L2TP offering here: https://www.aa.net.uk/broadband/l2tp-service/ and they happened to have to some guidance on setting it up with OpenWrt: https://support.aa.net.uk/L2TP_Client:_OpenWRT, so I gave it a try.
> 
> This uses a combination of PPP and DHCPv6. With mwan3 enabled is seems to causes issues with being able to get the IPv6 prefix delegated across the tunnel, when it's not enabled DHCPv6 works OK. It would seem mwan3 may be blocking the DHCPv6 related traffic somewhere.

@jamesmacwhite

There are many problems in using L2TP / PPTP and wireguard at the same time. I can only choose one of them now.

1f6c49c also has problems: when one of Wan and wanb interfaces is offline, L2TP / PPTP and wireguard will automatically connect to the online interface, which will cause the risk of abnormal IP of user terminal.

jamesmacwhite commented:
Potentially you could use static routes and metrics to make sure Wireguard or L2TP connects via a specific WAN to control that problem. I do exactly this, to ensure Wireguard/VPN traffic only goes through a specific WAN, rather my backup WAN which is data limited. I have noticed the routing table tends to sometimes get random routes across different WANs added, which can cause issues.

Having mwan3track use the default routing table and the source routing change would improve the situation a lot though. A lot of the time I think the problem with VPNs is due to the lack of gateway causing mwan3 to mark the interface as dead.

openwrt-diy commented:
> Potentially you could use static routes and metrics to make sure Wireguard or L2TP connects via a specific WAN to control that problem. I do exactly this, to ensure Wireguard/VPN traffic only goes through a specific WAN, rather my backup WAN which is data limited. I have noticed the routing table tends to sometimes get random routes across different WANs added, which can cause issues.
> 
> Having mwan3track use the default routing table and the source routing change would improve the situation a lot though. A lot of the time I think the problem with VPNs is due to the lack of gateway causing mwan3 to mark the interface as dead.

**I used the following static routing code without success**

`uci add network route 
uci set network.@route[-1].target='192.168.100.0/26'
uci set network.@route[-1].gateway='192.168.100.1'
uci set network.@route[-1].interface='wan'
uci set network.@route[-1].netmask='255.255.255.192'
uci commit network
uci add network route
uci set network.@route[-1].target='192.168.100.64/26'
uci set network.@route[-1].gateway='192.168.100.1'
uci set network.@route[-1].netmask='255.255.255.192'
uci set network.@route[-1].interface='wanb'
uci commit network
uci add network route
uci set network.@route[-1].target='192.168.100.128/26'
uci set network.@route[-1].gateway='192.168.100.1'
uci set network.@route[-1].netmask='255.255.255.192'
uci set network.@route[-1].interface='pptp'
uci commit network
uci add network route
uci set network.@route[-1].target='192.168.100.192/26'
uci set network.@route[-1].gateway='192.168.100.1'
uci set network.@route[-1].netmask='255.255.255.192'
uci set network.@route[-1].interface='l2tp'
uci commit network`

aaronjg commented:
> I recently experimented with L2TP and it does seem pretty broken currently even under 19.07.x. I believe the commit here to use default routing table helps the situation a lot.
> ...
> This uses a combination of PPP and DHCPv6. With mwan3 enabled is seems to causes issues with being able to get the IPv6 prefix delegated across the tunnel, when it's not enabled DHCPv6 works OK.

Interesting. mwan3 does have some rules to handle ipv6 ra packets: https://github.com/openwrt/packages/blob/3584eaa34ba9ca7ef4d9306ab8d3eab68df7c5d9/net/mwan3/files/lib/mwan3/mwan3.sh#L360
 and maybe some special handling is also needed for DHCPv6? If this issue was going on before this P/R, could you open a new issue and we can discuss it there?

----
>  There are many problems in using L2TP / PPTP and wireguard at the same time.
> 1f6c49c also has problems: when one of Wan and wanb interfaces is offline, L2TP / PPTP and wireguard will automatically connect to the online interface, which will cause the risk of abnormal IP of user terminal.

Yes, in general it can be tricky to set up tunnel interfaces with mwan3, since it will try to set them up over the currently active interface, so it may end up trying to route your L2TP traffic over wiregurad or vice-versa. You may need to set up a mwan3 rule for the wireguard or L2TP host to use the default routing table, and add a route to the default routing table directly out of the interface you want to the VPN endpoint (if the VPN netifd script did not already add it).

> 22ca579 also has a problem: when L2TP / PPTP interface is offline, mwan3 status page is displayed online!

Can you say a bit more about this? With the 1f6c49c and 22ca579, mwan3track is much better about sending traffic out of the right interface, so mwan3track may be able to `ping -I <iface>` when a normal ping would not. You can test pinging "as mwan3track would" using `DEVICE=<device> SRCIP=<src_ip> FWMARK=0x3f00 LD_PRELOAD=/lib/mwan3/libwrap_mwan3_sockopt.so.1.0 ping 1.1.1.1`

I did find a small bug in which ifdown events were not being properly communicated to mwan3track, and that is fixed here: 571a1e8. But even without that fix, if the interface was really down, then mwan3track would have marked it as offline when the pings failed.

openwrt-diy commented:
22ca579 also has a problem: when L2TP / PPTP interface is offline, mwan3 status page is displayed online!
![001](https://user-images.githubusercontent.com/57249132/91004595-3796a300-e607-11ea-8e83-1d6584ce34fc.png)
![002](https://user-images.githubusercontent.com/57249132/91004600-3b2a2a00-e607-11ea-9fb9-85af48c631eb.png)

22ca579 branch is a problem found before! I found that you fixed a lot of branches later. 
In order to save the time of compiling and testing, which branch do you choose to test now? I currently use two environmental scenarios:

2wan + 3vpn (PPTP L2TP wireguard) + mwan3 and 2wan + PPPoE server + mwan3 were compiled using openwrt 19.07-snapshot.

aaronjg commented:
Please test with the head commit on the branch. At the time of this comment: 09d9bee

In order to keep the number of commits that get merged into openwrt/packages:master down, I have rebased the changes into the previous commits, so while you may not see the number of commits change, I have updated the P/R.

Did `ifconfig <iface>` also show the interface as down?

openwrt-diy commented:
5e9138c feedback:

1. When adjusting the allocation strategy to the L2TP interface, you can access the Internet for a short time, and then there will be irregular online offline and redial! Finally, the Internet still cannot be accessed, and packets are dropped every time the ping is successful.

2. When adjusting the allocation strategy to the PPTP interface, you can access the Internet and ping the network within a short time. After that, the Internet cannot be accessed and the network cannot be pinged. The interesting thing is that the MWAN3 status page and network interface page show that PPTP is online, and the time is also displayed normally.

The WAN WANB WireGuard failover function is normal, and the IPv6 related interfaces have been deleted before the test.

aaronjg commented:
ec41fb9: fixed a procd issue that @jamesmacwhite found: https://github.com/openwrt/packages/issues/13195#issuecomment-680014368

aaronjg commented:
I added a function to the `/usr/sbin/mwan3` so that you can wrap a command in the helper library, binding it to a device and avoiding the mwan3 rules. This should be very helpful for testing out connections on the 5.x kernel and for testing interfaces with commands that do not support binding to a device.  Usage help added to the `/usr/sbin/mwan3` and in the commit message for fc14dfa393.

I also updated the helper library to work with `traceroute`, which made things a bit more complicated to avoid calling `bind` on an interface twice, however these changes should make the helper library more robust.

Additionally, now using `procd` to send the signals to the tracking interfaces from the hotplug script rather than finding the PID ourselves.

jamesmacwhite commented:
@aaronjg I like this. It makes testing multihoming stuff a bit easier, the iperf3 example being a recent case. Should save on having to create loads of rules with `option use_policy 'default'`.

aaronjg commented:
> @aaronjg I like this. It makes testing multihoming stuff a bit easier with mwan3 enabled, the iperf3 example being a recent case. Should save on having to create loads of rules with `option use_policy 'default'`.

Great! That's exactly the type of thing I wrote it for. Let me know if you have any issues.

It's still possible for some traffic to leave from other interfaces, for example dns lookups and reverse dns lookups done calling `glibc`'s `getaddrinfo` and `getnameinfo` but forcing these out of the the specified device could cause some unexpected behavior, and would add to code complexity, so I let those escape.

aaronjg commented:
52288da - fixed a race condition in mwan3track that happened when the up event was called before the down event was fully processed. Now ifup events will cancel pending ifdown events in mwan3track

aaronjg commented:
The circle-ci build was failing, with a problem building iproute2, which seemed unrelated to this P/R. It built fine on my system.  I am rebasing off of packages/master and I cleaned up the Makefile a bit, so hopefully that fixes the issue.

aaronjg commented:
replaced `sleep 2147483647` with a more understandable value "MAX_SLEEP=$(((1<<31)-1))"

Also added a "paused" status for mwan3track reporting, since with procd, all the mwan3track processes are started at once so it is helpful to know if the process has died (aka down), is running (aka active) or is running, but waiting for the interface to come online (aka paused)

aaronjg commented:
Fixed some race conditions identified by @wackejohn.

feckert commented on net/mwan3/Makefile:
```diff
@@ -26,6 +26,7 @@ define Package/mwan3
      +iptables-mod-conntrack-extra \
      +iptables-mod-ipopt \
      +jshn
+   PKG_CONFIG_DEPENDS:=IPV6
```
We need this for the C code, right?
If at all, then it must not be in the package definition but after the include in the top level. Also I think the vaule should be called CONFIG_IPV6.

aaronjg replied:
yes, it's needed to tell the build system to recompile if CONFIG_IPV6 is changed. You are right about the naming and location of this. will fix.

feckert commented on net/mwan3/Makefile:
```diff
@@ -61,8 +62,13 @@ fi
 endef
 
+define Build/Compile
+	$(TARGET_CC) $(CFLAGS) $(LDFLAGS) -fPIC -shared -o $(PKG_BUILD_DIR)/libwrap_mwan3_sockopt.so.1.0 $(if $(CONFIG_IPV6),-DCONFIG_IPV6) $(PKG_BUILD_DIR)/sockopt_wrap.c -ldl
```
The `-fPIC` should not get set this why. The right solution for this I can see in the package source is **$(FPIC)**

feckert commented on net/mwan3/files/etc/init.d/mwan3:
```diff
@@ -11,21 +14,47 @@ boot() {
 
 stop_service() {
-	/usr/sbin/mwan3 stop 1000>&-
+	/usr/sbin/mwan3 stop_
```
If we now switch to procd, wouldn't it be better to integrate the mwan3 tool into the procd service call? We can work with EXTRA_COMMAND so that we can add the additional calls for ifup/ifdown/wrap and so on. If someone wants to use the tool further, so we are backward compatible, we can change the mwan3 tool as a transitional solution so that it serves as a wrapper for the procd.

aaronjg replied:
Yes, this is an ugly hack. I was trying to make a smaller change, but you are right, this should all be in /etc/init.d/mwan3 and /usr/sbin/mwan3 should just be a light wrapper for that. I will fix this.

aaronjg replied:
I moved most of the work to the init.d script. I suppose we could go the whole distance and move the rest of the code from the sbin script, to the init.d script, export everything we need with EXTRA_COMMAND and then symlink /usr/sbin/mwan3 to /etc/init.d/mwan3.

aaronjg commented:
This is failing to build arc_arc700 build, however it appears that this issue is unrelated to the P/R, as many other P/Rs are failing on this test as well.

feckert commented on net/mwan3/files/lib/mwan3/mwan3.sh:
```diff
@@ -605,47 +586,48 @@ mwan3_add_all_nondefault_routes()
 mwan3_create_iface_rules()
 {
-	local id family IP
+	local id family IP i
```
Is this **i** a typo?

feckert commented on net/mwan3/files/lib/mwan3/mwan3.sh:
```diff
@@ -689,12 +666,8 @@ mwan3_delete_iface_rules()
-
-	while [ -n "$($IP rule list | awk '$1 == "'$(($id+2000)):'"')" ]; do
-		$IP rule del pref $(($id+2000))
+	for i in $(ip rule list | awk '$1 % 1000 == '$id' && $1 > 1000 && $1 < 4000 {print substr($1,0,4)}'); do
```
could you name the **i** to **rules** ?
On my vi you see **i** for insert and I sometimes don't know if this is a typo or not if I have a quick look on the code. Can you please change that.

feckert commented on net/mwan3/files/etc/hotplug.d/iface/15-mwan3:
```diff
@@ -7,9 +7,9 @@
+[ "$ACTION" = "ifup" ] || [ "$ACTION" = "ifdown" ] || [ "$ACTION" = "connected" ] || [ "$ACTION" = "disconnected" ] || exit 1
 [ -n "$INTERFACE" ] || exit 2
-if ( [ "$ACTION" = "ifup" ] || [ "$ACTION" = "connected" ] ) && [ -z "$DEVICE" ]; then
+if { [ "$ACTION" = "ifup" ] || [ "$ACTION" = "connected" ] ; } && [ -z "$DEVICE" ]; then
```
What is wrong with that?

aaronjg replied:
Parenthesis will fork and evaluate in a subshelll, whereas braces do not

neheb commented on net/mwan3/files/etc/hotplug.d/iface/15-mwan3:
```diff
@@ -7,18 +7,17 @@
 . /lib/mwan3/common.sh
```
Up a little bit, there's an include to functions.sh. I would remove and replace with /etc/rc.common in the shebang line.

aaronjg replied:
Why include /etc/rc.common in the shebang line? It seems like it is for init scripts, and not hotplug scripts. It does include the procd.sh but that is just for the procd_running and procd_send_signal functions.

neheb replied:
Ah ignore that.

feckert commented:
@aaronjg The way I see it we still have to fix the ubus call, because it currently does a false [pgrep](https://github.com/openwrt/packages/pull/13169/files#diff-d9589db021c8906200681f0926e02876R86). The check should look like this.
`pid="$(pgrep -f "mwan3track ${iface} )"` I get always a running false on ubus call, but that is not true.

@neheb Also why does arc_* does not build the LD_PRELOAD wrapper library?

feckert commented:
I am not sure which behavior is right now.
But I seem to remember that when I stop the mwan3 it triggers a disconnected event.
It does not do that anymore.
I think it is good that not all these events are triggered anymore.
But if the tracking is stopped then it should already trigger a disconnected event.
I also see that if I enter mwan3 ifdown <iface> then the the tragging is still in progress.
Should we not also disable the tracker and so start a disconnect event?

@aaronjg Any opinion on that? I would start a discussion on that.

aaronjg commented:
Hmm. It looks like there is some code duplication between the 'mwan3 interfaces' call and the ubus call, but these should really return the same results. Thanks for catching - will fix.

Good catch on the disconnect event. mwan3 triggers connected when it starts up, so it should trigger disconnected when it stops. I will fix that.

The build seems to fail because of iproute2, which had a version bump a few weeks ago.
https://bugs.openwrt.org/index.php?do=details&task_id=3340
https://bugs.openwrt.org/index.php?do=details&task_id=3353

jamesmacwhite reviewed (commented):
In addition to the check_quality options, I think once https://github.com/openwrt/luci/pull/4429 is merged, we can drop the `up`, `down` `timeout` and `interval` from the default configuration as well and let it fallback to mwan3track defaults.

Reliability needs to stay in because of this: https://github.com/openwrt/luci/issues/4428

Generally though interface definitions could be simplified in the default config to something like:

```
config interface 'wanX'
        option enabled '1'
	list track_ip '8.8.4.4'
	list track_ip '8.8.8.8'
	list track_ip '208.67.222.222'
	list track_ip '208.67.220.220'
        option reliability '1'
	option family 'ipv4'
```

Although keeping reliability of 2 on the main WAN is probably a good idea.

I was also thinking about adding this header to the config:

```
# Full documentation of options available: 
# https://openwrt.org/docs/guide-user/network/wan/multiwan/mwan3#mwan3_configuration
```

In the default configuration for those who configure without LuCI.

neheb commented:
@feckert I have no idea what you mean.

uClibc-ng is failing on iproute2. Updating it to version 1.0.35 fixes it. I have a patch on patchwork but nobody is merging.

aaronjg commented:
> @feckert I have no idea what you mean.
> 
> uClibc-ng is failing on iproute2. Updating it to version 1.0.35 fixes it. I have a patch on patchwork but nobody is merging.

The CI is reporting the following build failures, but it looks like it is failing to build  iproute2 and not even getting to trying to build the C code from mwan3. Maybe your fix will take care of it though.

> Test Build / arc_arc700 build (pull_request) Failing after 15m — arc_arc700 build
> Test Build / arc_archs build (pull_request) Failing after 15m — arc_archs build


@jamesmacwhite I think adding the comment is a great idea. That way users who aren't using LuCI will still know where to go to look for the additional options.

I had cherry-picked your commit on the configs, but I can drop that and then once the LuCI changes are in place you can make a single commit with the config changes.

feckert commented:
I'm thinking about creating a readme.md in the mwan3 directory and then documenting the configuration options here.
This would have the advantage that the options always fit to the branch if they are maintained.
But I think that would now lead too far in this P/R.
For the time being, I also think the proposal of @jamesmacwhite is good.

jamesmacwhite commented:
@aaronjg You can pull the commit from here: 

https://github.com/jamesmacwhite/openwrt-packages/tree/mwan3-default-config-options
https://github.com/jamesmacwhite/openwrt-packages/commit/ddabd91d0ed5e0b62976b074a16c34e41fa7ffc4

It removes most of the currently configured options and adds the comment header to the default config. Now LuCI has been harmonised with mwan3track most of them are redundant to have by default. I have left reliability in, even though the default is `1`, because of the default configuration using `option reliability '2'`, so just so it makes more sense for the wanb example. Although the LuCI warning is fixed now when not directly specifying it with this commit: https://github.com/openwrt/luci/commit/ab5b05dc2587323b6df971d9d6cf23844132571a

Cherry pick away!

feckert commented on net/mwan3/files/usr/libexec/rpcd/mwan3:
```diff
@@ -82,6 +82,7 @@ get_mwan3_status() {
 	network_get_device device $1
 
 	if [ "${iface}" = "${iface_select}" ] || [ "${iface_select}" = "" ]; then
+		#TODO use procd to see if tracker is running
```
So that it is not forgotten. Here the pgrep has to be adapted

aaronjg replied:
Thanks. I'm working on a fix now. I just need a few days to test for potential regressions.

feckert replied:
@aaronjg you delete in the last push force the pid but not the pgrep command in the next line. Could you also change this please.

aaronjg replied:
done

feckert commented:
Can please squash the commits https://github.com/openwrt/packages/pull/13169/commits/1328b3ef9c886fe794c257b78ff386451c9bf5ee and

aaronjg commented:
> Can you please squash the commits [1328b3e](https://github.com/openwrt/packages/commit/1328b3ef9c886fe794c257b78ff386451c9bf5ee) and [07f50aa](https://github.com/openwrt/packages/commit/07f50aaa05b3d6b7c33ea7bf441f1a5afcde4622) or give them a different commit message.

TY. Those should have been one commit. Squashed now.

feckert commented:
I have removed the WIP.
@aaronjg thanks for your hard work!
My comments have been implemented so far.
And my test seems ok.
So ready to merge?

aaronjg commented:
Yep! Sounds good. 👍

feckert commented:
@aaronjg I'm not available for the next four days, so I would not merge this until Tuesday.
If something does not work after merge I am not able to fix this, because this is a big change.
@James-TR @ptpt52 I also give the others more time to test it.

If everything is fine I will merge this on next Tuesday.

ptpt52 commented:
this is a big changes, if that so, I have to redo the dual stack support changes

aaronjg commented:
@feckert: makes sense. I believe you mean @jamesmacwhite? Also pinging @wackejohn and @wulfy23 for final testing

wackejohn commented:
@aaronjg 
Tested the latest commit ea95260, everything working fine except the ipv6 route expire problem, I tried modify the lib/mwan3.sh from 
`$IP route list table main | sed -ne "/^linkdown/T; s/expires \([0-9]\+\)sec/expires \1/;s/error [0-9]\+//; ${source_routing:+s/default\(.*\) from [^ ]*/default\1/;} p" | uniq`
to 
`$IP route list table main | sed -ne "/^linkdown/T; s/expire.*//;s/error [0-9]\+//; ${source_routing:+s/default\(.*\) from [^ ]*/default\1/;} p" | uniq`
(just delete the expire time when copy from the mail ipv6 routing table, and this fixes my problem.)

aaronjg commented:
Thanks @wackejohn . I just force pushed a fix for that. I also changed "mwan3 wrap" to "mwan3 use" per luizluca's suggestion (argument should tell what is being done, not how it is being done), and rebased off of master. 

CI appears to be broken at the moment with all current P/Rs failing Test Builds. However I did not change any of the C code between this and the last push, so this should not block.

feckert commented:
@luizluca regarding https://github.com/openwrt/packages/issues/13655#issuecomment-709807879 thanks for your detailed comments. But from my point of view I would like to follow the LD_PRELOAD approach. This has been in development and testing for 2 months now. Therefore I would like to add it to the mwan3 now, as it has advantages from my point of view.

feckert commented:
I thank @jamesmacwhite and @wackejohn for testing and especially @aaronjg for his contribution to the improvement of the mwan3 :+1:

luizluca commented:
> @luizluca regarding [#13655 (comment)](https://github.com/openwrt/packages/issues/13655#issuecomment-709807879) thanks for your detailed comments. But from my point of view I would like to follow the LD_PRELOAD approach. This has been in development and testing for 2 months now. Therefore I would like to add it to the mwan3 now, as it has advantages from my point of view.

That's fine, it's your call. Anyway, if anything does sideways, we have already a second strategy to try. Thanks for your hard work!

ptpt52 commented:
@aaronjg @feckert  I found some bug:

change `wan` interface track_ip from 1.1.1.1 to 1.2.3.4(expected it can not accessed)
and expected it go down but not.

ptpt52 commented:
in fact any changes on mwan3 interface track_ip (time count interval ... ) would not apply

ptpt52 commented:
and `mwan3 use` not work as expected

ptpt52 commented:
```
mwan3 use wan ping -c1 1.1.1.1
/usr/sbin/mwan3: line 155: use: not found
```

ptpt52 commented:
when mwan3track get ifup event it should reload all config from mwan3

aaronjg commented:
> when mwan3track get ifup event it should reload all config from mwan3

Chen, what is the use case for this? This behavior changed when we moved mwan3 to use OpenWRT's procd service framework. Other packages require "/etc/init.d/<service> [reload|restart]" to apply configuration changes. Changes to "/etc/config/mwan3" will cause flash writes, so should be minimized. Are you scripting changes to this? What is the objective? Perhaps there is another way that this can be done.

ptpt52 commented:
I just change config via web luci and not work as expected



aaronjg <notifications@github.com> 于 2020年10月22日周四 21:47写道：

> when mwan3track get ifup event it should reload all config from mwan3
>
> Chen, what is the use case for this? This behavior changed when we moved
> mwan3 to use OpenWRT's procd service framework. Other packages require
> "/etc/init.d/ [reload|restart]" to apply configuration changes. Changes to
> "/etc/config/mwan3" will cause flash writes, so should be minimized. Are
> you scripting changes to this? What is the objective? Perhaps there is
> another way that this can be done.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openwrt/packages/pull/13169#issuecomment-714506290>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABIVTYV2CUZU5Z6TA7KQNODSMAZYVANCNFSM4QE4SEZQ>
> .
>

feckert commented:
The configuration is actually not re-read when I change it and call /sbin/reload_config.
I think we have to extend the `/etc/init.d/mwan3` with the following code lines.

```
reload_service() {
       stop
       sleep 1
       start
}
```

aaronjg commented:
> I just change config via web luci and not work as expected

That's definitely a common use case :)

@feckert Yes, that appears to be correct. Potentially we could also be a bit smarter about the reload and send a signal to the mwan3track process, but let's try to get something in ASAP to fix this.

Is the `sleep 1` needed? I think the `stop` process will wait until all the procd processes are done before it returns.

ptpt52 commented:
but for now mwan3track get the ifup event it do not reload the mwan3 config

wulfy23 commented:
warning: abstract blabber herein...

init.d > /tmp/state < > *track >< ubus monitor || call reload_config ?

was pretty impressed with the mwan codebase while poking around... great work guys... it is a tad unusual with multiple procd processes within init.d and any watch|lock|state routines ...

above is a random observation from my perspective... ( from memory, lol, without properly looking at the code again... i.e. use state to have track talk to procd more directly ... or something along those lines anyway ), the goal being more of a 'flush/sync' than a stop-start... or at least a 'pre-stop' cleaner flush and dump... or something like that anyway...

2c

aaronjg commented:
> warning: abstract blabber herein...

Sorry, I'm not sure I follow.That command does not work for me. Could you explain your proposed solution?
